File: AdministracionDeTienda/src/gestorAplicacion/clasesHerencia/Proveedor.java
package gestorAplicacion.clasesHerencia;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Random;
import java.util.Set;

import gestorAplicacion.clasesBase.*;

public class Proveedor extends Persona{
	
	private float descuento;  
	private Bodega bodega;
	private int calle;
	private float costoCamiseta, costoPantalon, costoAbrigo;
	private float calificacion;
	
	public Proveedor(String nombre, int calificar, CuentaBancaria cuenta, Bodega bodega, float camiseta, float pantalon, float abrigo) {
		super(nombre,calificar,cuenta);
		this.bodega = bodega;
		costoCamiseta = camiseta;
		costoPantalon = pantalon;
		costoAbrigo = abrigo;
	}
	
	public void verificarDisponibilidad(ArrayList<Producto> pedido) {
		
		ArrayList<Producto> productosDisponibles = new ArrayList<Producto>();
		float descuentoCamiseta = 0;
		float descuentoPantalon = 0;
		float descuentoAbrigo = 0;
		int nCamisetas = 0;
		int nPantalones = 0;
		int nAbrigos = 0;
		
		for (int i=0; i< pedido.size();i++) {
			int aleatorio = (int)(Math.random()*5+1);
			
			if (aleatorio != 3) {
				if (pedido.get(i).getNombre().equals("CAMISA")) {
					nCamisetas++;
//					pedido.get(i).setCosto(costoCamiseta);
					if (aleatorio == 4) {descuentoCamiseta=descuentoCamiseta+2000f;}
					
				}
				else if(pedido.get(i).getNombre().equals("PANTALON")) {
					nPantalones++;
//					pedido.get(i).setCosto(costoPantalon);
					if (aleatorio == 4) {descuentoPantalon=descuentoPantalon+2500f;}
				}
				else {
//					pedido.get(i).setCosto(costoAbrigo);
					nAbrigos++;
					if (aleatorio == 4) {descuentoAbrigo=descuentoAbrigo+3000f;}
				}
				
				productosDisponibles.add(pedido.get(i));
				}
		}
		
		for (int i=0; i< productosDisponibles.size();i++) {
			if (productosDisponibles.get(i).getNombre().equals("CAMISA")) {
				productosDisponibles.get(i).setCosto(costoCamiseta-(descuentoCamiseta/nCamisetas));
				setCostoCamiseta(costoCamiseta-(descuentoCamiseta/nCamisetas));
			}
			else if(productosDisponibles.get(i).getNombre().equals("PANTALON")) {
				productosDisponibles.get(i).setCosto(costoPantalon-(descuentoPantalon/nPantalones));
				setCostoPantalon(costoPantalon-(descuentoPantalon/nPantalones));
			}
			else {
				productosDisponibles.get(i).setCosto(costoAbrigo-(descuentoAbrigo/nAbrigos));
				setCostoAbrigo(costoAbrigo-(descuentoAbrigo/nAbrigos));
			}
			this.descuento = descuentoCamiseta + descuentoPantalon + descuentoAbrigo;
		}
		
		bodega.setProductosEnBodega(productosDisponibles);
	}
	
	public  int calificar() {
		return 5;
	}

	public ArrayList<Producto> pDefectuososAReponer(ControlCalidad control) {
        ArrayList<Producto> productosAReponer = new ArrayList<>();
        ArrayList<Producto> listaProductos = control.getProductosDefectuosos();
        Random random = new Random();
        Set<Integer> indicesSeleccionados = new HashSet<>();

        int cantidadProductos = listaProductos.size();
        int cantidadSeleccionada = random.nextInt(cantidadProductos + 1);

        while (cantidadSeleccionada > 0) {
            int indiceAleatorio = random.nextInt(cantidadProductos);
            if (!indicesSeleccionados.contains(indiceAleatorio)) {
                Producto productoSeleccionado = listaProductos.get(indiceAleatorio);
                productosAReponer.add(productoSeleccionado);
                indicesSeleccionados.add(indiceAleatorio);
                cantidadSeleccionada--;
            }
        }

        return productosAReponer;
    }
	
	public void calificar(ControlCalidad c) {
        if (this != c.getProveedor()) {
            return;
        }
    
        ArrayList<Producto> productosAReponer = c.getProductosAReponerP();
        ArrayList<Producto> productosDefectuosos = c.getProductosDefectuosos();
        if (productosAReponer == null) {
            productosAReponer = new ArrayList<>();
        }
    
        int totalProductos = productosDefectuosos.size();
        int productosReemplazados = 0;
    
        for (Producto producto : productosDefectuosos) {
            if (productosAReponer.contains(producto)) {
                productosReemplazados++;
            }
        }
    
        int productosFaltantes = c.getProdsFaltantesCompra().size();
        int pedido = c.getCompra().getPedido().size();
    
        double porcentajeDeFaltanntes = ((double) productosFaltantes) / pedido;
        double porcentajeReemplazo = ((double) productosReemplazados) / totalProductos;
    
        int calificacion = this.calificar();
    
        if (productosDefectuosos.isEmpty() && productosAReponer.isEmpty() && porcentajeDeFaltanntes == 0) {
            this.setCalificacion(calificacion);
            return;
        } else if (productosDefectuosos.isEmpty() && productosAReponer.isEmpty() && porcentajeDeFaltanntes < 0.5) {
            this.setCalificacion(calificacion - 1);
            return;
        } else if (productosDefectuosos.isEmpty() && productosAReponer.isEmpty() && porcentajeDeFaltanntes >= 0.5) {
            this.setCalificacion(calificacion - 2);
            return;
        }
    
        if (productosReemplazados == totalProductos && porcentajeDeFaltanntes == 0) {
            this.setCalificacion(calificacion);
            return;
        } else if (productosReemplazados == totalProductos && porcentajeDeFaltanntes < 0.5) {
            this.setCalificacion(calificacion - 1);
            return;
        } else if (productosReemplazados == totalProductos && porcentajeDeFaltanntes >= 0.5) {
            this.setCalificacion(calificacion - 2);
            return;
        }
    
        if (porcentajeDeFaltanntes == 0) {
            if (porcentajeReemplazo >= 0.8) {
                this.setCalificacion(calificacion - 1);
            } else if (porcentajeReemplazo >= 0.6) {
                this.setCalificacion(calificacion - 2);
            } else if (porcentajeReemplazo > 0) {
                this.setCalificacion(calificacion - 3);
            } else {
                this.setCalificacion(calificacion - 4);
            }
            return;
        } else if (porcentajeDeFaltanntes < 0.5) {
            if (porcentajeReemplazo >= 0.8) {
                this.setCalificacion(calificacion - 2);
            } else if (porcentajeReemplazo >= 0.6) {
                this.setCalificacion(calificacion - 3);
            } else if (porcentajeReemplazo >= 0) {
                this.setCalificacion(calificacion - 4);
            }
            return;
        } else {
            if (porcentajeReemplazo >= 0.8) {
                this.setCalificacion(calificacion - 3);
            } else {
                this.setCalificacion(calificacion - 4);
            }
            return;
        }
    }
	

	public void setCostoCamiseta(float costo) { costoCamiseta = costo;}
	
	public void setCostoPantalon(float costo) { costoPantalon = costo;}
	
	public void setCostoAbrigo(float costo) { costoAbrigo = costo;}
	
	public float getCostoCamiseta() { return costoCamiseta;}
	
	public float getCostoPantalon() { return costoPantalon;}
	
	public float getCostoAbrigo() { return costoAbrigo;}
	
	public float getDescuento() {return descuento;}
	
	public Bodega getBodega() {return bodega;}
	
	public int getCalle() {return calle;}
	
	public float costoProductos() {
		return bodega.calcularCamisas() * getCostoCamiseta()+ bodega.calcularPantalon()*getCostoPantalon()+ bodega.calcularAbrigos()*costoAbrigo;	
	}
	
	
	public String toString() {
	  return "camisetas: "+bodega.calcularCamisas(getBodega().getProductosEnBodega())+ " X " +getCostoCamiseta() + "\nPantaloes: " + bodega.calcularPantalon(getBodega().getProductosEnBodega()) + " X " + getCostoPantalon() + "\nAbrigos: " + bodega.calcularAbrigos(getBodega().getProductosEnBodega()) + " X " + getCostoAbrigo();
		
	}

}

File: Python/src/gestorAplicacion/clasesBase/Proveedor.py
from Persona import *
from random import randint
import random
import math


class Proveedor(Persona):
    def __init__(self, nombre, calificacion, cuenta, bodega, camisa, pantalon, abrigo, calle):
        super().__init__(nombre, calificacion, cuenta)
        self.bodega = bodega
        self.costoCamisa = camisa
        self.costoPantalon = pantalon
        self.costoAbrigo = abrigo
        self.calle = calle
        self.descuento = None


    def verificarDisponibilidad(self, pedido):
        descuentoCamiseta = 0
        descuentoPantalon = 0
        descuentoAbrigo = 0
        numeroCamisetas = 0
        numeroPantalon = 0
        numeroAbrigo = 0
        productosDisponibles = []

    #De forma aleatoria se decide que productos tiene disponible el proveedor
    #De forma aleatoria se decide si el proveedor ofrece descuento por los productos

        for producto in pedido:
            aleatorio = randint(0, 6)

            if aleatorio != 3:
                productosDisponibles += [producto]

                if producto.getNombre() == "CAMISA":
                    numeroCamisetas += 1
                    if aleatorio == 4 or aleatorio == 5:
                        descuentoCamiseta += producto.getCosto()*0.20

                elif producto.getNombre() == "PANTALON":
                    numeroPantalon += 1
                    if aleatorio == 4 or aleatorio == 5:
                        descuentoPantalon += producto.getCosto() * 0.20

                else:
                    numeroAbrigo += 1
                    if aleatorio == 4 or aleatorio == 5:
                        descuentoAbrigo += producto.getCosto() * 0.20

        #Se le asigna el precio a cada producto en funcion del costo

        for producto1 in productosDisponibles:
            if producto1.getNombre() == "CAMISA":
                producto1.setCosto(round(self.costoCamisa - descuentoCamiseta / numeroCamisetas, 2))

            elif producto1.getNombre() == "PANTALON":
                producto1.setCosto(round(self.costoPantalon - descuentoPantalon / numeroPantalon,2))

            else:
                producto1.setCosto(round(self.costoAbrigo - descuentoAbrigo / numeroAbrigo, 2))

            producto1.setPrecio(round(producto1.getCosto()*2, 2))

        self.bodega.setProductosEnBodega(productosDisponibles)
        self.setDescuento(round(descuentoPantalon + descuentoCamiseta + descuentoAbrigo, 2))

    def pDefectuososAReponer(self, control):
        productosSeleccionados = []
        indicesSeleccionados = set()
        listaProductos = control.getProductosDefectuosos()

        cantidadProductos = len(listaProductos)
        cantidadSeleccionada = random.randint(0, cantidadProductos)

        while cantidadSeleccionada > 0:
            indiceAleatorio = random.randint(0, cantidadProductos - 1)
            if indiceAleatorio not in indicesSeleccionados:
                productoSeleccionado = listaProductos[indiceAleatorio]
                productosSeleccionados.append(productoSeleccionado)
                indicesSeleccionados.add(indiceAleatorio)
                cantidadSeleccionada -= 1
        return productosSeleccionados


    def calificar(self, c):
        if self != c.getProveedor():
            return

        productosAReponer = c.getProductosAReponerP()
        productosDefectuosos = c.getProductosDefectuosos()
        if productosAReponer is None:
            productosAReponer = []

        totalProductos = len(productosDefectuosos)
        productosReemplazados = 0

        for producto in productosDefectuosos:
            if producto in productosAReponer:
                productosReemplazados += 1

        productosFaltantes = len(c.getProdsFaltantesCompra())
        pedido = len(c.getCompra().getPedido())

        porcentajeDeFaltanntes = productosFaltantes / pedido

        if totalProductos > 0:
            porcentajeReemplazo = productosReemplazados / totalProductos
        else:
            porcentajeReemplazo = 0.0

        calificacion = self.valorCalificacion()

        if productosDefectuosos == [] and productosAReponer == [] and porcentajeDeFaltanntes == 0:
            self.setCalificacion(calificacion)
            return
        elif productosDefectuosos == [] and productosAReponer == [] and porcentajeDeFaltanntes < 0.5:
            self.setCalificacion(calificacion - 1)
            return
        elif productosDefectuosos == [] and productosAReponer == [] and porcentajeDeFaltanntes >= 0.5:
            self.setCalificacion(calificacion - 2)
            return

        if productosReemplazados == totalProductos and porcentajeDeFaltanntes == 0:
            self.setCalificacion(calificacion)
            return
        elif productosReemplazados == totalProductos and porcentajeDeFaltanntes < 0.5:
            self.setCalificacion(calificacion - 1)
            return
        elif productosReemplazados == totalProductos and porcentajeDeFaltanntes >= 0.5:
            self.setCalificacion(calificacion - 2)
            return

        if porcentajeDeFaltanntes == 0:
            if porcentajeReemplazo >= 0.8:
                self.setCalificacion(calificacion - 1)
            elif porcentajeReemplazo >= 0.6:
                self.setCalificacion(calificacion - 2)
            elif porcentajeReemplazo > 0:
                self.setCalificacion(calificacion - 3)
            else:
                self.setCalificacion(calificacion - 4)
            return
        elif porcentajeDeFaltanntes < 0.5:
            if porcentajeReemplazo >= 0.8:
                self.setCalificacion(calificacion - 2)
            elif porcentajeReemplazo >= 0.6:
                self.setCalificacion(calificacion - 3)
            elif porcentajeReemplazo >= 0:
                self.setCalificacion(calificacion - 4)
            return
        else:
            if porcentajeReemplazo >= 0.8:
                self.setCalificacion(calificacion - 3)
            else:
                self.setCalificacion(calificacion - 4)
            return
            
    def getCuenta(self):
        return self.cuenta

    def setCuenta(self, cuenta):
        self.cuenta = cuenta
    
    def valorCalificacion(self):
        return 5

    def setDescuento(self, descuento):
        self.descuento = descuento

    def getDescuento(self):
        return self.descuento

    def getCalle(self):
        return self.calle

    def setCalle(self, calle):
        self.calle = calle

    def getBodega(self):
        return self.bodega

    def setBodega(self, bodega):
        self.bodega = bodega






